pkg/bgp/speaker/events.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package speaker
+
+import (
+	"go.universe.tf/metallb/pkg/k8s/types"
+	metallbspr "go.universe.tf/metallb/pkg/speaker"
+)
+
+type svcEvent struct {
+	name string
+	svc  *metallbspr.Service
+	eps  *metallbspr.Endpoints
+}
+type epEvent svcEvent
+
+// nodeEvent must be a pointer to a map because maps are not hashable, aka
+// cannot be keys in a map, which queue is.
+type nodeEvent *map[string]string
+
+// run runs the reconciliation loop, fetching events off of the queue to
+// process. The events supported are svcEvent, epEvent, and nodeEvent. This
+// loop is only stopped (implicitly) when the Agent is shutting down.
+//
+// Adapted from go.universe.tf/metallb/pkg/k8s/k8s.go.
+func (s *Speaker) run() {
+	for {
+		key, quit := s.queue.Get()
+		if quit {
+			return
+		}
+		st := s.do(key)
+		switch st {
+		case types.SyncStateError:
+			s.queue.AddRateLimited(key)
+		case types.SyncStateSuccess, types.SyncStateReprocessAll:
+			// SyncStateReprocessAll is returned in MetalLB when the
+			// configuration changes. However, we are not watching for
+			// configuration changes because our configuration is static and
+			// loaded once at Cilium start time.
+			s.queue.Forget(key)
+		}
+	}
+}
+
+// do performs the appropriate action depending on the event type. For example,
+// if it is a service event (svcEvent), then it will call into MetalLB's
+// SetService() to perform BGP announcements.
+func (s *Speaker) do(key interface{}) types.SyncState {
+	defer s.queue.Done(key)
+
+	switch k := key.(type) {
+	case svcEvent:
+		e := key.(svcEvent)
```

Comment:
there's no need to do this, line `64` already has the converted type, so in here `k` is already `svcEvent`. Same applies for line 69 and 72

Reply:
You're right. For some reason, I thought I would get compilation errors if I didn't, but turns out there's no need.